### ISSUE TITLE
dnn unable to train ggplot::diamonds

### ISSUE THREAD

**Issue by tripartio** [citoverse/cito#123]
Hello. I'm having trouble using `cito` on a pretty standard dataset: `ggplot2::diamonds`. No matter what I try, the intermediate loss functions are all `NA` and the final predictions are all `NaN`. Here's what I've tried:

``` r
library(cito)

diamonds <- ggplot2::diamonds

# Default settings:
nn <- dnn(price ~ ., diamonds)
#> Warning: the 'nobars' function has moved to the reformulas package. Please update your imports, or ask an upstream package maintainter to do so.
#> This warning is displayed once per session.
#> Loss is NA. Bad training, please adjust hyperparameters. See vignette('B-Training_neural_networks') for help.
# Loss is NA. Bad training, please adjust hyperparameters. See vignette('B-Training_neural_networks') for help.
predict(nn, diamonds) |> head()
#>   [,1]
#> 1  NaN
#> 2  NaN
#> 3  NaN
#> 4  NaN
#> 5  NaN
#> 6  NaN

# Try numeric variables only
nn <- dnn(
  price ~ .,
  diamonds |> dplyr::select((where(is.numeric)))
)
#> Loss is NA. Bad training, please adjust hyperparameters. See vignette('B-Training_neural_networks') for help.
predict(nn, diamonds) |> head()
#>   [,1]
#> 1  NaN
#> 2  NaN
#> 3  NaN
#> 4  NaN
#> 5  NaN
#> 6  NaN


# Try tips from https://citoverse.github.io/cito/articles/B-Training_neural_networks.html

# Increase epochs and tune learning rate
nn <- dnn(
  price ~ .,
  data = diamonds,
  epochs = 200,
  lr = tune(c(0.001, 0.1)),
)
#> Starting hyperparameter tuning...
#> Fitting final model...
predict(nn, diamonds) |> head()
#>   [,1]
#> 1  NaN
#> 2  NaN
#> 3  NaN
#> 4  NaN
#> 5  NaN
#> 6  NaN

# Use learning rate scheduler
nn <- dnn(
  price ~ .,
  data = diamonds,
  epochs = 200,
  lr = 0.2,
  lr_scheduler = config_lr_scheduler("reduce_on_plateau", patience = 5, factor = 0.5),
)
#> Loss is NA. Bad training, please adjust hyperparameters. See vignette('B-Training_neural_networks') for help.
predict(nn, diamonds) |> head()
#>   [,1]
#> 1  NaN
#> 2  NaN
#> 3  NaN
#> 4  NaN
#> 5  NaN
#> 6  NaN
```

<sup>Created on 2026-03-07 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

I also tried almost randomly tuning lots of things. It was very slow but still didn't work. Here's the code, though I don't show the results--they were identical to all the tuned options above (`NaN` for all predictions).

``` r
nn <- dnn(
  price ~ .,
  data = diamonds,
  hidden = tune(),
  epochs = tune(),
  batchsize = tune(),
  lr = tune(),
)
predict(nn, diamonds) |> head()
```

Could you please help me figure out what I'm doing wrong with this very standard dataset?

**Comment by MaximilianPi** [comment]
Hi @tripartio 

There are two problems:
1) The scale of your response and your variables:
Huge scales are more challenging for the gradient optimizer and require tuning of the learning rate. I suggest scaling them. You can scale the predictions back, it shouldn't affect the predictive performance.

2) Ordered factors:
Currently, cito's internal formula handling cannot deal with ordered factors because the orthogonal polynomial contrasts do not match the names of the levels. However, the orthogonal contrasts do not make sense because they are a way to check the functional form of the trends through the ordered factors. The DNN can change the linear trend/contrast into a nonlinear effect. I will add a checker for ordered factors that will automatically transform them to integers in the development version.

Here is code that should work without the development version:
```

diamonds <- ggplot2::diamonds

diamonds_transformed = 
  diamonds |> mutate(price = price/max(price), 
                     depth = scale(depth)[,1], 
                     table = scale(table)[,1], 
                     cut = as.integer(cut), 
                     color = as.integer(color), 
                     clarity = as.integer(clarity))

# Default settings:
nn <- dnn(price ~ ., diamonds_transformed)
```

**Comment by tripartio** [comment]
Thanks for this. With a few tests in `cito 1.1.1`, I found that all I need is to scale the numeric variables (including the target outcome `price`):

``` r
library(cito)
library(dplyr)

diamonds <- ggplot2::diamonds

diamonds_transformed <- diamonds |> 
  mutate(across(
    where(is.numeric),
    \(it.col) scale(it.col)[ , 1]
  ))

nn <- dnn(price ~ ., diamonds_transformed)
#> Warning: the 'nobars' function has moved to the reformulas package. Please update your imports, or ask an upstream package maintainter to do so.
#> This warning is displayed once per session.
#> Loss at epoch 1: 0.724561, lr: 0.01000
#> Loss at epoch 2: 0.278895, lr: 0.01000
#> ...
#> Loss at epoch 100: 0.069913, lr: 0.01000
```

![](https://i.imgur.com/BVdQA3u.png)


<sup>Created on 2026-03-11 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Perhaps you could add an argument to give the option to scale some or all numeric variables.

In any case, thanks again; my issue is resolved.
